main/src/ca/uwaterloo/flix/language/phase/EarlyTreeShaker.scala

```diff
@@ -0,0 +1,388 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.CompilationMessage
+import ca.uwaterloo.flix.language.ast.Ast.Annotation
+import ca.uwaterloo.flix.language.ast.TypedAst._
+import ca.uwaterloo.flix.language.ast.{Symbol, TypedAst}
+import ca.uwaterloo.flix.util.Validation._
+import ca.uwaterloo.flix.util.{InternalCompilerException, ParOps, Validation}
+
+/**
+  * The Tree Shaking phase removes all unused function definitions.
+  *
+  * A function is considered reachable if it:
+  *
+  * (a) The main function is always reachable.
+  * (b) A function marked with @benchmark or @test is reachable.
+  * (c) Appears in a function which itself is reachable.
+  *
+  */
+object EarlyTreeShaker {
+
+  /**
+    * Performs tree shaking on the given AST `root`.
+    */
+  def run(root: Root)(implicit flix: Flix): Validation[Root, CompilationMessage] = flix.phase("EarlyTreeShaker") {
+    // Compute the symbols that are always reachable.
+    val initReach = initReachable(root)
+
+    // Compute the symbols that are transitively reachable.
+    val allReachable = ParOps.parReachable(initReach, visitSym(_, root))
+
+    // Filter the reachable definitions.
+    val reachableDefs = root.defs.filter {
+      case (sym, _) => allReachable.contains(ReachableSym.DefnSym(sym))
+    }
+
+    // Reassemble the AST.
+    root.copy(defs = reachableDefs).toSuccess
+  }
+
+  /**
+    * Returns the symbols that are always reachable.
+    */
+  private def initReachable(root: Root): Set[ReachableSym] = {
+    // A set used to collect the symbols of reachable functions.
+    var reachable: Set[ReachableSym] = Set.empty
+
+    //
+    // (a) The main function is always reachable (if it exists).
+    //
+    reachable = reachable ++ root.entryPoint.toList.map(ReachableSym.DefnSym)
+
+    //
+    // (b) A function annotated with @benchmark or @test is always reachable.
+    //
+    for ((sym, defn) <- root.defs) {
+      if (isBenchmark(defn) || isTest(defn)) {
+        reachable = reachable + ReachableSym.DefnSym(sym)
+      }
+    }
+    reachable
+  }
+
+  /**
+    * Returns `true` if `defn` is annotated with `@benchmark`
+    */
+  private def isBenchmark(defn: TypedAst.Def): Boolean = defn.spec.ann.exists { a =>
+    a.name match {
+      case Annotation.Benchmark(_) => true
+      case _ => false
+    }
+  }
+
+  /**
+    * Returns `true` if `defn` is annotated with `@test`
+    */
+  private def isTest(defn: TypedAst.Def): Boolean = defn.spec.ann.exists { a =>
+    a.name match {
+      case Annotation.Test(_) => true
+      case _ => false
+    }
+  }
+
+  /**
+    * Returns the symbols reachable from the given symbol `sym`.
+    */
```

Comment:
Explain what you include here. E.g.

Every expression in a class instance of a reachable class symbol is reachable.